src/cuda/api/memory.hpp

```diff
@@ -221,6 +222,117 @@ inline void copy(void *destination, const void *source, size_t num_bytes)
+{
+	cudaMemcpy3DParms copyParams = {0};
+
+	// make_cudaPitchedPtr expects a void* pointer
+	void* source_ = const_cast<void*>(source);
+
+	// How to create a pitched ptr from a `const void*` ?
+	copyParams.srcPtr = make_cudaPitchedPtr(
+	    source_, destination.dims()[0] * sizeof(T),  destination.dims()[0], destination.dims()[1]);
+
+	copyParams.dstArray = destination.get();
+
+	cudaExtent ext = make_cudaExtent(destination.dims()[0], destination.dims()[1], destination.dims()[2]);
+
+	copyParams.extent = ext;
+
+	copyParams.kind = cudaMemcpyDefault;
+
+	auto result = cudaMemcpy3D(&copyParams);
+	if (is_failure(result)) {
+		std::string error_message("Synchronously copying into array");
+		throw_if_error(result, error_message);
+	}
+}
+
+/**
+ * Synchronously copies data between memory spaces and CUDA arrays.
+ *
+ * @param destination A pointer to a a memory region of size `source.size() * sizeof(T)`
+ * @param source A CUDA array @ref cuda::array::array_t
+ */
+template<typename T>
+void copy(void* destination, const array::array_t<T, 3>& source)
+{
+	cudaMemcpy3DParms copyParams = {0};
+
+	copyParams.dstPtr = make_cudaPitchedPtr(
+	    destination, source.dims()[0] * sizeof(T),  source.dims()[0], source.dims()[1]);
+
+	copyParams.srcArray = source.get();
+
+	cudaExtent ext = make_cudaExtent(source.dims()[0], source.dims()[1], source.dims()[2]);
+
+	copyParams.extent = ext;
+
+	copyParams.kind = cudaMemcpyDefault;
+
+	auto result = cudaMemcpy3D(&copyParams);
+	if (is_failure(result)) {
+		std::string error_message("Synchronously copying from array");
+		throw_if_error(result, error_message);
+	}
+}
+
+/**
+ * Synchronously copies data between memory spaces and CUDA arrays.
+ *
+ * @param destination A CUDA array @ref cuda::array::array_t
+ * @param source A pointer to a a memory region of size `destination.size() * sizeof(T)`
+ */
+template<typename T>
+void copy(const array::array_t<T, 2>& destination, const void *source)
+{
+	// Consider the padded array:
+	// 
+	// x x x x o o o
+	// x x x x o o o
+	// x x x x o o o
+	// 
+	// o = padded element
+	// x = actually used element of the array
+	// 
+	// The pitch in the example above is 7 * sizeof(T)
+	// The width is 4 * sizeof(T)
+	// The height is 3
+	// 
+	// See also https://stackoverflow.com/questions/16119943/how-and-when-should-i-use-pitched-pointer-with-the-cuda-api
+	// 
+	auto result = cudaMemcpy2DToArray(destination.get(), 0, 0, source, destination.dims()[0] * sizeof(T), destination.dims()[0] * sizeof(T), destination.dims()[1], cudaMemcpyDefault);
+
+	if (is_failure(result)) {
+		std::string error_message("Synchronously copying into array");
+		throw_if_error(result, error_message);
+	}
+}
+
+/**
+ * Synchronously copies data between memory spaces and CUDA arrays.
+ *
+ * @param destination A pointer to a a memory region of size `source.size() * sizeof(T)`
+ * @param source A CUDA array @ref cuda::array::array_t
+ */
+template<typename T>
```

Comment:
If the array_t class id templated on the number of dimensions, are you sure this function shouldn't be templated either? Again, with definitions only of specializations for 2D and 3D? 

I don't have a strong opinion on this.

Reply:
Letting this go for now, will revisit the question at a later time.